### ISSUE TITLE
apply_inference_nonlin after averaging probabilities

### ISSUE THREAD

**Issue by TimZaragori** [MIC-DKFZ/nnUNet#2781]
Hi, 
I am trying to understand how nnUNet ensembling is performed and I'm wondering about the use of `label_manager.convert_logits_to_segmentation` which internally calls `self.apply_inference_nonlin` after averaging the probabilities.
Here's what I came up with:

- When we call `nnUNetv2_predict` with `--save_probabilities` inference non-linearity layer is applied and what is stored in the `.npz` files are probabilities and not logits
- Probabilities coming from different models are averaged
- `label_manager.convert_logits_to_segmentation` is called instead of `label_manager.convert_probabilities_to_segmentation` because after averaging in the multiclass case, you can have the sum of the probabilities for a voxel that is not 1. To ensure this, you apply once again the inference non-linearity layer. 

Can you confirm that this is the right thinking?
In a binary case, reapplying the sigmoid layer doesn't change the results ?

**Comment by TimZaragori** [comment]
Since I've dug a little deeper into this part of the code, I think there is a mistake in `merge_files` when saving the averaged probabilities. Actually we have:
```
  if save_probabilities:
      np.savez_compressed(output_filename_truncated + '.npz', probabilities=probabilities)
      save_pickle(probabilities, output_filename_truncated + '.pkl')
```
However, shouldn't the saved pickle file be `properties` instead of `probabilities`? which give this:
```
  if save_probabilities:
      np.savez_compressed(output_filename_truncated + '.npz', probabilities=probabilities)
      save_pickle(properties, output_filename_truncated + '.pkl')
```

**Comment by TimZaragori** [comment]
Any updates on this ?

**Comment by Karol-G** [comment]
Hey,

Is this issue still relevant? Otherwise, I will close it in 14 days.

Best,
Karol

**Comment by TimZaragori** [comment]
Hi, I didn't get any answers on my interrogations (potential things to correct depending on the answers):

- `label_manager.convert_logits_to_segmentation` [link](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/ensembling/ensemble.py#L43) which internally calls `self.apply_inference_nonlin` after averaging the probabilities. I understand this use in the multiclass case, but I am wondering if this is approriate in binary case with a sigmoid layer
- There might be a typo when saving the averaged results [link](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/ensembling/ensemble.py#L45)

Best,
Timothée

**Comment by FabianIsensee** [comment]
Hey Tim, 
too bad I am only seeing your issue now. The mistake in merge_files where convert_logits instead of convert_probabilities was used has been there for a long time and I would have liked to fix it a long time ago. 
I stumbled upon that by mere accident a couple weeks ago and resolved it in commit d5a1437dc7cd13d714dabbe88dc59ad31b56b73a
The pickle saving still receives the wrong input, thanks for pointing this out. If you want to create a PR for that (so that finding and fixing that gives you the credit you deserve) then please feel free to do so. If not I would just fix it myself - your call :-)
Best,
Fabian

**Comment by FabianIsensee** [comment]
Hey @TimZaragori can you please tell me whether you plan to submit your own PR or would like me to fix this? Thanks!

**Comment by TimZaragori** [comment]
Hi @FabianIsensee I just did it, sorry for the delay

**Comment by FabianIsensee** [comment]
Thanks!
